### PR TITLE
Collision rebased

### PR DESCRIPTION
**Pull request by MrChico**
based on new-gas-branch

### CODE REVIEW

**Comment by MrChico** [comment]
To review this PR, you will need to take https://github.com/dapphub/k/pull/4 and https://github.com/dapphub/evm-semantics/pull/3 into account. I'll summarize the changes here:
evm storage is modeled with an abstract `IMap: Int -> Int`, where entries are of the form `store(M, K, V)`, where `M : IMap`, and `K`, `V` are integers representing a key and a value. To fetch a storage location, one uses `select(M, K)`, which reduces according to the rules:
```
    rule select(store(M, K0, V), K) => V            requires K0  ==Int K
    rule select(store(M, K0, V), K) => select(M, K) requires K0 =/=Int K
```
Besides this, I have added rules simplifying `stores` with provably equal entries: https://github.com/dapphub/evm-semantics/pull/3/files#diff-494b4c66d4128fe395a8c9607fab3629R886, and lemmas asserting injectivity of `keccakIntList`.
There is also a new cell in the configuration, `<origStorage>`, somewhat unrelated, but incorporated in newer version of `evm-semantics` to make it easier to incorporate some new EIP I believe. The changes in klab are to accomodate for the new configuration format and throwing a warning if a possible storage collision is sensed. 

Will break the current k-dss specs. Working on a PR for this as well

File: resources/run.sh [hunk 1]
```diff
@@ -1,2 +1,16 @@
...
-K_OPTS=-Xmx10G $KLAB_EVMS_PATH/.build/k/k-distribution/target/release/k/bin/kprove --debugg --debugg-path $TMPDIR/klab --debugg-id $2 --directory $KLAB_EVMS_PATH/.build/java/ --z3-executable --def-module RULES --output-tokenize "<k>" --output-omit "<programBytes> <program> <code> <previousGas> <touchedAccounts> <interimStates> <callStack>" --output-flatten "_Map_ #And" --output json --smt_prelude ./prelude.smt2 --z3-tactic "(or-else (using-params smt :random-seed 3 :timeout 1000) (using-params smt :random-seed 2 :timeout 2000) (using-params smt :random-seed 1))" $1
+K_OPTS=-Xmx10G \
+$KLAB_EVMS_PATH/.build/k/k-distribution/target/release/k/bin/kprove \
+--debugg \
+--debugg-path $TMPDIR/klab \
+--debugg-id $2 \
+--directory $KLAB_EVMS_PATH/.build/java/ \
+--z3-executable \
+--def-module RULES \
+--output-tokenize "#And _==K_ <k> #unsigned <storage>" \
```
Comment (mhhf) [inline resources/run.sh]: we don't tokenize most things anymore - this is needed for levs gas-solver

File: lib/build.js [hunk 2]
```diff
@@ -370,12 +370,23 @@ const buildAct = config => ({act, oog, pass, name}) => {
     "ethereum.network.accounts":       accounts,
   })
     const buildAccount = varname => {
-      let fstore = Object.keys(act.storage[varname])
+        let storage_pass = "_"
+        Object.keys(act.storage[varname])
+            .forEach(key => {
```
Comment (mhhf) [inline lib/build.js]: plz use avoid forEach and use map/reduce whenever possible

**Comment by mhhf** [comment]
@livnev can you review the lemma/smt changes?

**Comment by livnev** [comment]
@MrChico Hey sorry for the delay I types all of this up a couple of days ago and forgot to press send.

Working on reviewing/understanding this now. First question: it seems like the new way of composing map lookups as 
```
store(store(M:IMap, K1, V1), K2, V2)
```
instead of 
```
K1 |-> V1
K2 |-> V2
_:Map
```
is possibly non-commutative. This means that if we are compiling a sequence of storage lookups from an act spec, we need to make sure that whatever we have compiled is provably commutative, since the semantics of the storage in the act specs is assumed commutative. so that's why we would need to throw if we determined that there was a possibly key collision in the spec. I'll think about what else we may need to do to guarantee that our language is sound, if anything.

Could use some clarification on [this lemma](https://github.com/dapphub/evm-semantics/pull/3/files#r239639200).

Also I think I managed to convince @dbrock that smart contract languages should forbid you to refer to potentially colliding storage keys, and that the correct behaviour of `move(src, dst, wad)` is to `REVERT` when `src == dst`. Now it remains only to convince @rainbreak :smiling_imp:

**Comment by MrChico** [comment]
It is true that the map is non-commutative when keys are provably equal, 
`store(store(M:IMap, K1, V1), K2, V2)` and `store(store(M:IMap, K2, V2), K1, V1)` will reduce to `store(store(M:IMap, K1, V2)` and the second `store(store(M:IMap, K1, V2)` if K can deduce `K1 == K2`. However, the same is true for the current map:
```
    #hashedLocation("Solidity", 0, To)        |-> ToBal
    #hashedLocation("Solidity", 0, CALLER_ID) |-> FromBal
```
reduces to 
```
    #hashedLocation("Solidity", 0, CALLER_ID) |-> FromBal
```
while 
```
    #hashedLocation("Solidity", 0, CALLER_ID) |-> FromBal
    #hashedLocation("Solidity", 0, To)        |-> ToBal
```
reduces to 
```
    #hashedLocation("Solidity", 0, CALLER_ID) |-> ToBal
```
the changes include a warning if klab can detect a possible collision (only for entries that are defined along the lines of https://github.com/dapphub/k-dss/blob/master/src/storage.k.md though), but we can change that to throw an error instead

**Comment by livnev** [comment]
@MrChico Yes, I understand how it's true of the current map: the problem there arose also out of assumed commutativity.

When we switch to `IMap`: if `klab` misses a possible collision (I'm calling these "storage keys in generic/special position"), the consequence is that the proof should fail right? Or is there a chance that the proof for something like `move` will still succeed if the keys happen to be in the right order and the actual rewrites end up being sound?

**Comment by MrChico** [comment]
@livnev 
The problem this solves isn't really non-commutativity of the current map structure, rather that entries are implicitly assumed distinct. This structure does solves that issue, and yes, if there is a possible collision in two entries then the proof will fail. What I we are discussing above is a problem it doesn't solve, which is to make sense of (obviously non-sensical specs) where multiple entries under the `storage` header have provably equal keys. 

The reason why this store is better is in the side condition of select; unless you can prove that `K1 =/= K2` in `select(store(M, K1, V1), K2, V2)`, the expression will remain symbolic, and yes, a proof like `move` would fail, since it tries to `select` entries that are not provably distinct. 

However, you do bring up a good point, I can think of a spec that (I think) would succeed even though it should be false (or at least non-sensical) for the case `K1 == K2`:
```sol
contract Token{
  mapping(address => uint) public balanceOf;
  function santa(address nice, uint256 gift) public {
    balanceOf[nice] = add(balanceOf[to], gift);
  }
}
```act
behaviour santa of Token
interface santa(address nice, uint gift)

types

    NiceBal          : uint256
    NaughtyBal   : uint256
    Naughty        : address

storage

    #hashedLocation("Solidity", 0, Naughty) |-> NaughtyBal
    #hashedLocation("Solidity", 0, Nice)        |-> NiceBal   => NiceBal + Value


iff in range uint256

    NiceBal + Value

```

**Comment by livnev** [comment]
@MrChico :heart: the christmas spec!

Yeah that's what I was vaguely concerned about. In the example you gave, whether or not it succeeds might depend on the order of the keys in the `storage` block, right? Since it won't be able to commute the nested `store` since it doesn't know that `Naughty =/= Nice`.

We still won't be able to prove anything "definitely false" though right? The only issue is that we might allow a spec to slip through that we don't really want to accept as a semantically valid spec, but that we will still be proving "what the user probably meant"? Or do you see some way in which this can be more unsound (I realise that once you accept these "special position" specs it's already philosophically unsound in some way)?

Another thing:

> What I we are discussing above is a problem it doesn't solve, which is to make sense of (obviously non-sensical specs) where multiple entries under the storage header have provably equal keys.

When you say "have provably equal keys" do you mean "don't have provably distinct keys"?

**Comment by MrChico** [comment]
I suppose we could enforce some policy like:
```
rule store(M, K1, V1) => #ERRORSTORE
    requires notBool #isDistinct(M, K1)
   andBool notBool #inKeys(M, K1)
```
where `#isDistinct(M, K1)` is only true if `K1` is provably distinct from all keys of `M`

**Comment by MrChico** [comment]
Argh, we would if `K` would not assume stupid `LEM`

**Comment by livnev** [comment]
@MrChico That seems pretty sensible to me.

**Comment by livnev** [comment]
@MrChico could you clarify [this](https://github.com/dapphub/evm-semantics/pull/3/files#r239639200) as well please?

**Comment by MrChico** [comment]
@livnev you did see my reply? Can you tell me what can be made clearer

**Comment by livnev** [comment]
> @livnev you did see my reply? Can you tell me what can be made clearer

@MrChico I can't find your reply to that question, no. :thinking:

**Comment by MrChico** [comment]
screen shot 2018-12-09 at 23 15 58

**Comment by livnev** [comment]
@MrChico My plan is to now to make the appropriate changes to `k-dss` and run against this PR to see if everything works as expected.

Are there any others parts of this PR I should focus on?

**Comment by MrChico** [comment]
@livnev This has been updated with the new conservative storage rules. I moved the IMap into evm-semantics instead of domains.k until we get some clarity on this https://github.com/kframework/k/pull/208 so now the relevant PRs are https://github.com/dapphub/k/pull/5 and https://github.com/dapphub/evm-semantics/pull/3

**Comment by livnev** [comment]
@MrChico Please rebase this onto `dev`.

**Comment by MrChico** [comment]
ci fails for the same reason it does on `dev`, but as everything passed ci before rebase and I just tested `Token_transfer_pass_rough.k` and `Token_transfer_fail.k` locally and it works, I'm pretty confident that this works and would like to merge this into dev. Please review @mhhf @livnev

File: lib/build.js [hunk 3]
```diff
@@ -412,14 +418,25 @@ const buildAct = config => ({act, oog, pass, name}) => {
   if (pass) {
       gasCond = "VGas " + (oog ? "<Int " : ">=Int ") + act.gas
   }
+
+  //The multiplication of the list monad  
+  const mu = ListOfList => ListOfList.reduce((a, list) => a.concat(list))
```
Comment (mhhf) [inline lib/build.js]: this is usually called "flatten"

Comment (mhhf) [reply lib/build.js]: also don't giving reduce an initial value (e.g. `[]`) will result in a crash whenever ListOfList is empty (e.g. `[]`)

File: lib/storage.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+const mapInterface = (is, str) => {
+  is.forEach(([t, n]) => {
+    str = str.replace(new RegExp("([^\\w\.]|^)" + n + "([^\\w]|$)", "g"), (a, b, c, d) => {
+      return b + "ABI_" + n + c;
+    })
+  })
+  return str;
+}
+
+const toK = str => str
```
Comment (mhhf) [inline lib/storage.js]: this is redundant with lib/build.js
can we somehow unify them in either utils.js or kast.js or another one?!

File: lib/build.js [hunk 5]
```diff
@@ -341,10 +342,9 @@ const parseAct = (act_str, silent = false) => {
   return act;
 }
 
-
 const buildAct = config => ({act, oog, pass, name}) => {
-  console.log("build " + name)
-
+  console.log("building " + name)
```
Comment (mhhf) [inline lib/build.js]: "build" plz

Comment (mhhf) [reply lib/build.js]: cause "parse" is 5 letters as well :P

Comment (MrChico) [reply lib/build.js]: Sounds like imperative form to me... Why is klab telling me to `build`? I think `building`, `parsing`, is a lot clearer.

**Comment by livnev** [comment]
@MrChico I'm happy with the functionality and the K stuff, perhaps @mhhf wants to sign off on the javascript since I didn't review that.

**Comment by MrChico** [comment]
@mhhf addressed your points and would like to merge this. I will work on a pretty printing of the storage, but that shouldn't stop this from getting into `dev`

**Pull request merged**